oauth/client_factory.go

```diff
@@ -14,156 +14,139 @@
+	"errors"
 	"fmt"
+	"io"
+	"net/http"
 	"net/url"
 
-	"github.com/redhat-appstudio/service-provider-integration-operator/api/v1beta1"
-	authz "k8s.io/api/authorization/v1"
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/meta"
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/runtime/schema"
+	"github.com/codeready-toolchain/api/api/v1alpha1"
 	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
-	certutil "k8s.io/client-go/util/cert"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-// AuthenticatingClient is just a typedef that advertises that it is safe to use the WithAuthIntoContext or
-// WithAuthFromRequestIntoContext functions with clients having this type.
-type AuthenticatingClient client.Client
+const wsApiPath = "/apis/toolchain.dev.openshift.com/v1alpha1/workspaces"
 
-type ClientFactoryConfig struct {
-	KubeInsecureTLS bool
-	KubeConfig      string
-	ApiServer       string
-	ApiServerCAPath string
+var (
+	errUnableToParseWorkspaceResponse = errors.New("unable to parse response from workspace API requested for namespace")
+	errWorkspaceNotFound              = errors.New("target workspace not found for namespace")
+	errClientInstanceNotSet           = errors.New("client instance does not set")
+)
+
+type K8sClientFactory interface {
+	CreateClient(ctx context.Context) (client.Client, error)
 }
 
-// CreateUserAuthClient creates a new client based on the provided configuration. We use this client for k8s requests with user's token (e.g.: SelfSubjectAccessReview).
-// Note that configuration is potentially modified during the call.
-func CreateUserAuthClient(args *ClientFactoryConfig) (AuthenticatingClient, error) {
-	// we can't use the default dynamic rest mapper, because we don't have a token that would enable us to connect
-	// to the cluster just yet. Therefore, we need to list all the resources that we are ever going to query using our
-	// client here thus making the mapper not reach out to the target cluster at all.
-	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
-	mapper.Add(authz.SchemeGroupVersion.WithKind("SelfSubjectAccessReview"), meta.RESTScopeRoot)
-	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessToken"), meta.RESTScopeNamespace)
-	mapper.Add(v1beta1.GroupVersion.WithKind("SPIAccessTokenDataUpdate"), meta.RESTScopeNamespace)
-
-	return createClient(args, mapper, true)
+type InClusterK8sClientFactory struct {
+	ClientOptions *client.Options
+	RestConfig    *rest.Config
 }
 
-// CreateInClusterClient creates a new client based on the provided configuration. We use this client for k8s requests with ServiceAccount (e.g.: reading configuration secrets).
-func CreateInClusterClient(args *ClientFactoryConfig) (client.Client, error) {
-	mapper := meta.NewDefaultRESTMapper([]schema.GroupVersion{})
-	mapper.Add(corev1.SchemeGroupVersion.WithKind("Secret"), meta.RESTScopeNamespace)
+type WorkspaceAwareK8sClientFactory struct {
```

Comment:
can you please document what is it doing? Workspaces are not common Kubernetes feature, so it's really useful to have some docs around.

Reply:
ok